aopy/precondition.py

```diff
@@ -1,2 +1,161 @@
 # precondition.py
-# code for preconditioning neural data
\ No newline at end of file
+# code for preconditioning neural data
+from scipy import signal
+from scipy.signal import butter, lfilter, filtfilt, windows, freqz
+import math
+import numpy as np
+import os
+import matplotlib.pyplot as plt
+#
+import nitime.algorithms as tsa
+import nitime.utils as utils
+from nitime.viz import winspect
+from nitime.viz import plot_spectral_estimate
+'''
+Filter functions
+'''
+def generate_test_signal(T, fs, freq, a):
+    '''
+    Generates a test time series signal with multiple frequencies, specified in freq, for T timelength at a sampling rate of fs
+
+    Args:
+        T (float): Time period in seconds
+        fs (int): Sampling frequency in Hz
+        freq (1D array): Frequencies to be mixed in the test signal. main frequency in the first element
+        a (1D array) : amplitudes for each frequencies and last element of the array to be amplitude of noise (size : len(freq) + 1)
+
+    Returns:
+        x (1D array): cosine wave with multiple frequencies and noise
+        t (1D array): time vector for x
+    '''
+    nsamples = int(T * fs)
+    t = np.linspace(0, T, nsamples, endpoint=False)
+    # a = 0.02
+    f0 = freq[0]
+    # noise_power = 0.001 * fs / 2
+    x = a[-1] * np.sin(2 * np.pi * 1.2 * np.sqrt(t))  # noise
+    # x += np.random.normal(scale=np.sqrt(noise_power), size=t.shape)  # noise
+
+    for i in range(len(freq)):
+        x += a[i] * np.cos(2 * np.pi * freq[i] * t)
+
+    return x, t
+
+def butterworth_params(cutoff_low, cutoff_high, fs, order = 4, filter_type = 'bandpass'):
+    '''
+    Design Nth-order digital Butterworth filter and return the filter coefficients.
+
+    Args:
+        cutoff_low (int): lower cut-off frequency (in Hz)
+        cutoff_high (int): higher cutoff frequency (in Hz)
+        fs (int): sampling rate (in Hz)
+        order (int): Order of the butter worth filter
+        filter_type (str) : Type of filter. Accepts one of the four values - {‘lowpass’, ‘highpass’, ‘bandpass’, ‘bandstop’}
+
+    Returns:
+        tuple (b,a): bandpass filter parameters
+
+    '''
+    b,a = butter( order, [cutoff_low, cutoff_high], btype=filter_type, fs =fs)
+    return b,a
+
+def butterworth_filter_data(data, cutoff_low, cutoff_high, fs, order = 4,filter_type = 'bandpass' ):
```

Comment:
also would remove filter_type from here and rename it to butterworth_bandpass_filter_data(), since it won't work if you only pass cutoff_low or vice versa.